receive_stream.go

```diff
@@ -121,21 +122,29 @@ func (s *receiveStream) readImpl(p []byte) (bool /*stream completed */, int, err
 				return false, bytesRead, s.resetRemotelyErr
 			}
 
-			if !s.deadline.IsZero() && !time.Now().Before(s.deadline) {
-				return false, bytesRead, errDeadline
+			deadline := s.deadline
+			if !deadline.IsZero() {
+				if !time.Now().Before(deadline) {
```

Comment:
I'm guessing this is an optimization, not an actual check? Maybe you should add a comment that says so.

Reply:
Why do you think this is an optimization?

Reply:
Well, it's either that or a race condition, no? The code has to deal with the case of the deadline being less than now one instruction after this check, so it can't rely on the check for correctness.

Reply:
It’s inside the mutex so there’s no race here. Also, this check is in the for loop, so this is a necessary way to exit that loop after the deadline expires, not an optimization.

Reply:
Oh right, I missed the for loop 🙄LGTM